patient_data_source.py

```diff
@@ -0,0 +1,169 @@
+
+class Patient(ABC):
+
+	@abstractmethod
+	def get_gender(self) -> str:
+		"""Returns the patient's gender."""		
+		pass
+
+	@abstractmethod
+	def get_id(self) -> str:
+		"""Returns the patient's id."""
+		pass
+
+	@abstractmethod
+	def get_dob(self) -> datetime:
+		"""Returns the patient's DOB."""
+		pass
+
+	def generate_name(self, gender):
+		"""Returns a first and last name based on gender."""
+		return names.get_last_name(), names.get_first_name('male' if gender=='M' else 'female')
+
+	def get_name(self):
+		return self.generate_name(self.get_gender())
+
+class Observation(ABC):
+	class ObservationType(Enum):
+		fio2 = 0
+		pip = 1
+		peep = 2
+		hr = 3
+		sao2 = 4
+
+	# Map unit strings from the VALUEUOM column of CHARTEVETS to codes that follow the spec in
+    # https://ucum.org/ucum.html
+	UNIT_CODES = {
+	  ObservationType.fio2 : '',
+	  ObservationType.pip : 'cm[H20]',
+	  ObservationType.peep : 'cm[H20]',
+	  ObservationType.hr : '/min',
+	  ObservationType.sao2 : '%',
+	}
+
+	# LOINC codes that I found by using loinc.org/search/ ... and making my best guesses when things were unclear
+	# Not to be fully trusted
+	LOINC_CODES = {
+	  ObservationType.fio2 : '19996-8',
+	  ObservationType.pip : '60951-1',
+	  ObservationType.peep : '20077-4',
+	  ObservationType.hr : '8867-4',
+	  ObservationType.sao2 : '59408-5',
+	}
+
+	DISPLAY_STRINGS = {
+	  ObservationType.fio2 : 'ETT Sx Quality',
+	  ObservationType.pip : 'PIP',
+	  ObservationType.peep : 'PEEP',
+	  ObservationType.hr : 'Heart Rate',
+	  ObservationType.sao2 : 'Sa02',
+	}
+
+	@abstractmethod
+	def get_id(self) -> str:
+		"""Returns the observation's id."""
+		pass
+
+	@abstractmethod
+	def get_observation_type(self) -> ObservationType:
+		pass
+	
+	def get_unit_string(self) -> str:
+		return self.UNIT_CODE[self.get_observation_type()]
+
+	def get_display_string(self) -> str:
+		return self.DISPLAY_STRINGS[self.get_observation_type()]
+
+	def get_unit_code(self) -> str:
+		return self.UNIT_CODE[self.get_observation_type()]
+
+	def get_loinc_code(self) -> str:
+		return self.LOINC_CODES[self.get_observation_type()]
+
+	@abstractmethod
+	def get_value(self) -> str:
+		pass
+
+	@abstractmethod
+	def get_time(self) -> datetime:
+		"""Returns the patient's name."""
+		pass
+
+class PatientDataSource(ABC):
+	"""Abstract class for loading patient data into a smart FHIR server"""
+	
+	@abstractmethod
+	def get_all_patients(self):
+		"""Returns a list of all Patients."""
+		pass
+
+	@abstractmethod
+	def get_patient_observations(self, patient : Patient):
+		"""Returns a list of all Observation for one patient."""
+		pass
+
+	@abstractmethod
+	def get_patient_system(self):
+		"""Returns the system where the patient information is stored."""		
+		pass
+
+	def create_patient(self, patient: Patient) -> FHIR_Patient :
+		"""Create a smart FHIR_Patient object."""
+		gender = patient.get_gender()
+		family, given = patient.get_name()
+		return FHIR_Patient({
```

Comment:
Let's include the patient gender in here too. But only if `patient.get_gender` was implemented. One way to address conditional implementation is to make the default implementation return `None`, and then do nothing with it if you see the `None` here